HACKING.md

```diff
@@ -311,43 +322,61 @@ To make a release of the ops library, do the following:
+5. Use the "Generate Release Notes" button to get a copy of the changes into the
    notes field.
-5. Group the changes by the commit type (feat, fix, etc.) and use full names (e.g., "Features",
-   not "feat") for group headings. Strip the commit type prefix from the bullet point. Strip the
-   username (who did each commit) if the author is a member of the Charm Tech team.
-6. Where appropriate, collapse multiple tightly related bullet points into a
+6. Group the changes first by package (`ops` and `ops-scenario`) and then by
+   the commit type (feat, fix, etc.) and use full names (e.g., "Features",
+   not "feat") for group headings. Strip the commit type prefix from the bullet point,
+   and capitalise the first word. Strip the username (who did each commit) if the
+   author is a member of the Charm Tech team.
+   For example: the PR `docs: clarify where StoredState is stored`
+   becomes `* Clarify where StoredState is stored` in the "Documentation" section.
+7. Where appropriate, collapse multiple tightly related bullet points into a
    single point that refers to multiple commits.
-7. Create a new branch, and copy this text to the [CHANGES.md](CHANGES.md) file,
+8. Create a new branch, and copy this text to the [CHANGES.md](CHANGES.md) file,
    stripping out links, who did each commit, the new contributor list, and the
    link to the full changelog.
-8. Change [version.py](ops/version.py)'s `version` to the
+9. Change [version.py](ops/version.py)'s `version` to the
    [appropriate string](https://semver.org/).
-9. Check if there's a `chore: update charm pins` auto-generated PR in the queue. If it looks
+10. Check if there's a `chore: update charm pins` auto-generated PR in the queue. If it looks
    good, merge it and check that tests still pass. If needed, you can re-trigger the
    `Update Charm Pins` workflow manually to ensure latest charms and ops get tested.
-10. Add, commit, and push, and open a PR to get the changelog and version bump
+11. Add, commit, and push, and open a PR to get the changelog and version bump
    into main (and get it merged).
-11. Back in the GitHub releases page, tweak the release notes - for example,
+12. Back in the GitHub releases page, tweak the release notes - for example,
    you might want to have a short paragraph at the intro on particularly
    noteworthy changes.
-12. Have someone else in the Charm-Tech team proofread the release notes.
-13. When you are ready, click "Publish". (If you are not ready, click "Save as Draft".)
-
-This will trigger an automatic build for the Python package and publish it to
-[PyPI](https://pypi.org/project/ops/)) (authorisation is handled via a
+13. Have someone else in the Charm-Tech team proofread the release notes.
+14. When you are ready, click "Publish". (If you are not ready, click "Save as Draft".)
+15. If the release includes an `ops-scenario` package, then create a new tag in
+   the form `scenario-<major>.<minor>.<patch>`.
```

Comment:
How to create a (second) new tag? I'm hoping there's a way in the github web ui, but I can't figure it out. Normally I just tag stuff via the cli, but I don't think that's what we want here since it would require being able to push directly to ops (main).

Reply:
Hmm, yes, this is a problem. I hadn't thought about that - too used to having high access to a repo. Maybe we do have to have three types of tags after all? Or maybe we simplify things by just using one tag and always doing two releases, or the release workflow tries to be clever and see if the number has changed.

I'll need to think more about this.

Reply:
My gut feeling is that just making two separate releases with `2.x.y` or `scenario-a.b.c` tags respectively would be the way to go, but if there does exist an easy way to do what you originally proposed then that would be better.

Reply:
Maybe that would be the cleanest way of working in the system. The disadvantages would be that there are two sets of source artefacts (but I don't think anyone uses those anyway), the person doing the release would need to do a few things twice, and the release notes wouldn't be able to refer to both releases (although the Discourse post is probably the more true release notes, and could).

Reply:
> The disadvantages would be that there are two sets of source artefacts
A bit redundant, but totally fine and correct imo

> the person doing the release would need to do a few things twice
Biggest downside! But maybe the duplicated work isn't too much? You'd have to separate out the commits between ops and opt-testing anyway, for example.

> the release notes wouldn't be able to refer to both releases
Which release notes? In github you would still start with auto generated release notes vs correct previous release tag and separate out the (non)testing ones (for example), right? A single discourse post when releasing both would make sense though.

Just my 2c

Reply:
> Which release notes? In github you would still start with auto generated release notes vs correct previous release tag and separate out the (non)testing ones (for example), right?

Yes. I meant that say there's some significant change that impacts both ops and ops-scenario: if there was one combined GitHub release, then that could have some note about it at the top, once, and it's probably one PR, so it would be one bullet point in the list below. With two GitHub releases, you need to call it out at the top of each of them, and the bullet point is in both of the lists.

Even for a little change this seems a bit annoying: most features are going to also include corresponding changes in Scenario. For example:

* Adds support for the new workload-ready Juju event (#19790125)

This ends up being in the GitHub release for `ops` and for `ops-scenario`, because for `ops` it's the feature itself, but then there's the support for testing it in `ops-scenario`.

It's not a blocker, it just means that it's not super straightforward to filter the lists for the two releases, particularly if you didn't work on or review the PR yourself.

Reply:
Oh yeah I didn't think of that, thanks for explaining. Kind of puts me off the idea of two truly separate releases.

Probably a bad idea: when we release both, we could just create a second dummy release for testing with a link to the ops release with combined release notes (with separate sections where appropriate). So basically your originally proposed workflow, but the 'create a second tag' step is done via creating a second release for the same commit.